src/go/k8s/internal/testutils/k3s.go

```diff
@@ -0,0 +1,605 @@
+	discoveryv1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	serviceLabel = "kubernetes.io/service-name"
+
+	defaultTimeout          = 1 * time.Minute
+	defaultK3sImage         = "rancher/k3s:v1.27.1-k3s1"
+	defaultHelmChart        = "redpandadata/redpanda"
+	defaultHelmRepo         = "redpandadata"
+	defaultHelmURL          = "https://charts.redpanda.com"
+	defaultCertManagerChart = "jetstack/cert-manager"
+	defaultCertManagerRepo  = "jetstack"
+	defaultCertManagerURL   = "https://charts.jetstack.io"
+)
+
+var defaultPorts = map[string]int{
+	"admin":          9644,
+	"http":           8082,
+	"kafka":          9093,
+	"rpc":            33145,
+	"schemaRegistry": 8081,
+}
+
+type K3sStackOptions struct {
+	K3sImage            string
+	CertManagerURL      string
+	CertManagerRepo     string
+	CertManagerChart    string
+	CertManagerVersion  string
+	Driver              string
+	InstallationTimeout time.Duration
+}
+
+func (k K3sStackOptions) normalize() K3sStackOptions {
+	if k.K3sImage == "" {
+		k.K3sImage = defaultK3sImage
+	}
+
+	if k.CertManagerURL == "" {
+		k.CertManagerURL = defaultCertManagerURL
+	}
+
+	if k.CertManagerRepo == "" {
+		k.CertManagerRepo = defaultCertManagerRepo
+	}
+
+	if k.CertManagerChart == "" {
+		k.CertManagerChart = defaultCertManagerChart
+	}
+
+	if k.InstallationTimeout == 0 {
+		k.InstallationTimeout = defaultTimeout
+	}
+
+	return k
+}
+
+type RedpandaOptions struct {
+	Name        string
+	Namespace   string
+	HelmURL     string
+	HelmRepo    string
+	HelmChart   string
+	HelmVersion string
+	Values      map[string]interface{}
+}
+
+func (r RedpandaOptions) normalize(suffix int32) RedpandaOptions {
+	defaultNameAndNamespace := fmt.Sprintf("k3s-%d-%d", time.Now().Unix(), suffix)
+
+	if r.Name == "" {
+		r.Name = defaultNameAndNamespace
+	}
+
+	if r.Namespace == "" {
+		r.Namespace = defaultNameAndNamespace
+	}
+
+	if r.HelmURL == "" {
+		r.HelmURL = defaultHelmURL
+	}
+
+	if r.HelmRepo == "" {
+		r.HelmRepo = defaultHelmRepo
+	}
+
+	if r.HelmChart == "" {
+		r.HelmChart = defaultHelmChart
+	}
+
+	if r.Values == nil {
+		r.Values = map[string]interface{}{}
+	}
+
+	return r
+}
+
+type K3sStack struct {
+	Options       K3sStackOptions
+	config        []byte
+	helm          helmclient.Client
+	directory     string
+	client        client.Client
+	rest          *rest.Config
+	installations atomic.Int32
+	clusters      map[int32]*v1alpha2.Redpanda
+	portForwarder *portForwarder
+	mutex         sync.RWMutex
+
+	ports     []*forwardedPorts
+	portMutex sync.RWMutex
+}
+
+func RunK3sStack(t *testing.T, options K3sStackOptions) *K3sStack {
```

Comment:
I didn't realize that testcontainers had a `k3s` module (TBH I've not used testcontainers). I added a [`k3d`](https://k3d.io/) [package](https://github.com/redpanda-data/redpanda-operator/blob/main/src/go/k8s/pkg/k3d/k3d.go) last week because we needed the ability to simulate node failures.

How was the experience of test containers?

Reply:
I also have a vague interest in playing around with [vcluster](https://www.vcluster.com/docs/vcluster/configure/vcluster-yaml/) as a way to get full cluster isolation without dedicating so many resources to running multiple clusters.